lib/client/redirect.go

```diff
@@ -336,12 +344,20 @@ func (rd *Redirector) wrapCallback(fn func(http.ResponseWriter, *http.Request) (
 			case rd.errorC <- err:
 			case <-rd.context.Done():
 			}
-			http.Redirect(w, r, errorURL, http.StatusFound)
+			redirectURL := errorURL
+			if keys.IsPrivateKeyPolicyError(err) {
+				redirectURL = terminalRedirectURL
+			}
+			http.Redirect(w, r, redirectURL, http.StatusFound)
 			return
 		}
 		select {
 		case rd.responseC <- response:
-			http.Redirect(w, r, successURL, http.StatusFound)
+			redirectURL := successURL
+			if rd.RequiresHardwareKey {
+				redirectURL = terminalRedirectURL
+			}
```

Comment:
With `hardware_key_pin`, `tsh` promts for PIN after just the first popup, so we can redirect to the success screen here. Same with `hardware_key` which doesn't require any interaction.
```suggestion
			switch rd.PrivateKeyPolicy {
			case PrivateKeyPolicyHardwareKey:
				// login should complete without user interaction, success.
			case keys.PrivateKeyPolicyHardwareKeyPIN:
				// The user is prompted to enter their PIN before this step,
				// so we can go straight to success screen.
			case keys.PrivateKeyPolicyHardwareKeyTouch, keys.PrivateKeyPolicyHardwareKeyTouchAndPIN:
				// The user is prompted for touch their hardware key after
				// this redirect, so display the terminal redirect screen.
				redirectURL = terminalRedirectURL
			}
```